spark/v3.4/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/MergeRowsExec.scala

```diff
@@ -66,115 +67,219 @@ case class MergeRowsExec(
+        if (action.cond.eval(row)) {
+          return action.apply(row)
+        }
+      }
 
-    def processRowWithCardinalityCheck(rowIdOrdinal: Int)(inputRow: InternalRow): InternalRow = {
-      val isSourceRowPresent = isSourceRowPresentPred.eval(inputRow)
-      val isTargetRowPresent = isTargetRowPresentPred.eval(inputRow)
+      if (emitNotMatchedTargetRows) targetTableProj.apply(row) else null
+    }
 
-      if (isSourceRowPresent && isTargetRowPresent) {
-        val currentRowId = inputRow.getLong(rowIdOrdinal)
-        if (matchedRowIds.contains(currentRowId)) {
-          throw new SparkException(
-            "The ON search condition of the MERGE statement matched a single row from " +
-            "the target table with multiple rows of the source table. This could result " +
-            "in the target row being operated on more than once with an update or delete " +
-            "operation and is not allowed.")
+    private def applyNotMatchedActions(row: InternalRow): InternalRow = {
+      for (action <- notMatchedActions) {
+        if (action.cond.eval(row)) {
+          return action.apply(row)
         }
-        matchedRowIds.add(currentRowId)
       }
 
-      if (emitNotMatchedTargetRows && !isSourceRowPresent) {
-        projectTargetCols.apply(inputRow)
-      } else if (!isTargetRowPresent) {
-        applyProjection(nonMatchedPairs, inputRow)
+      null
+    }
+  }
+
+  class SplittingMergeRowIterator(
+      private val rowIterator: Iterator[InternalRow],
+      private val cardinalityCheck: CardinalityCheck,
+      private val isTargetRowPresentPred: BasePredicate,
+      private val matchedActions: Seq[Action],
+      private val notMatchedActions: Seq[Project])
+    extends Iterator[InternalRow] {
+
+    var cachedExtraRow: InternalRow = _
+
+    override def hasNext: Boolean = cachedExtraRow != null || rowIterator.hasNext
+
+    override def next(): InternalRow = {
+      if (cachedExtraRow != null) {
+        val extraRow = cachedExtraRow
+        cachedExtraRow = null
+        return extraRow
+      }
+
+      val row = rowIterator.next()
+
+      // it should be OK to just check if the target row exists
+      // as this iterator is only used for delta-based row-level plans
+      // that are rewritten using an inner or right outer join
+      if (isTargetRowPresentPred.eval(row)) {
+        cardinalityCheck.execute(row)
+        applyMatchedActions(row)
       } else {
-        applyProjection(matchedPairs, inputRow)
+        applyNotMatchedActions(row)
       }
     }
 
-    val processFunc: InternalRow => InternalRow = if (performCardinalityCheck) {
-      val rowIdOrdinal = child.output.indexWhere(attr => conf.resolver(attr.name, ROW_ID))
-      assert(rowIdOrdinal != -1, "Cannot find row ID attr")
-      processRowWithCardinalityCheck(rowIdOrdinal)
-    } else {
-      processRow
+    private def applyMatchedActions(row: InternalRow): InternalRow = {
+      for (action <- matchedActions) {
```

Comment:
Feel free to ignore this if it's too Scala ish but I would do something like
```scala
return matchedActions.find(row => action.cond.eval(row)).match{
   case split: Split => 
     cachedExtraRow = split.projectExtraRow(row)
     split.projectRow(row)
   case project: Project =>
      project.apply(row)
  }.orElse(null)
  ```
  
  For these find first situations, I would also probably just keep the option and return None but that's not required.

Reply:
This is something that is invoked per record so I wanted it to produce as few objects as possible and as simple bytecode as possible, hoping JIT would then make smart choices.

I do like None and find but I am paranoid it would add more calls.

Reply:
I think that's a good paranoia :) I just think it's a little less readable. I have no problem keeping it this way but you do already have benchmarks set up so you could test it... ;)

Reply:
Let me make a note of this and test it in parallel to working on distribution and ordering.